Support other visualizers not included in the default set

richyen opened the pull request:
Gives users the ability to use `--target custom` and provide a `--post_url` and `--plan_key` to hook into a visualizer not in the default set (depesz/dalibo/tensor).  A use case for this would be an infrastructure where a clone of explain.depesz source was deployed in a private LAN.

richyen committed:
Add support for explain.tensor.ru

richyen committed:
merge with upstream fork

richyen committed:
Refactor code to remove code-duplication
This consolidates visualizer targets into the visualizer package
This will eventually make way for the possibility of passing in
`--target=custom`, along with `--post_url` and `--plan_key`, so
users can post their plans to local installs of say, depesz
(see https://gitlab.com/depesz/explain.depesz.com/-/blob/master/Vagrantfile
for an example of how someone might deploy their own
explain.depesz.com instance)

richyen committed:
golint changes

richyen committed:
remove default, to allow pass-through to error case; default is established in main.go flag.String()

richyen committed:
Support other visualizers not included in the default set

agneum commented on README.md:
```diff
@@ -77,8 +78,10 @@ On default, make install puts the compiled binary in `go/bin`.
   - `tensor` - https://explain.tensor.ru
+  - `custom` - Any visualizer that you wish to use.  If using a `custom` target, you will need to provdie the following:
+    - `--post_url` - The absolute URL to which the `<form>` will `POST` to.  A good reference would be to look for the `action` param in the `<form>` tag.  Note that `plan-exporter` only works with visualizers supporting HTTP `POST` at this time.
+    - `--plan_key` - The `name` of the `<textarea>` (or other input) element into which the query plan gets pasted
```
I suggest adding only the `post_url` parameter. In my opinion, it would be better to explicitly implement the new type of goal, rather than modify the `plan_key` option.

richyen replied:
I agree, but how do you plan to get the `plan_key`? Would we write an HTML scraper to fetch the HTML element, or something like that (or maybe there's something more elegant that I'm not seeing)?

agneum replied:
We can simply add a new package to the visualizer, if necessary.
Unlike `post_url`, there are few visualization services, and they can be explicitly added as code.
However, `post_url` should be customizable.

richyen replied:
Agreed -- there are not that many visualizer services out there, so it's not going to be too much trouble to create new packages as needed.  In that case, it may make more sense to abandon this PR and add packages via PR as needed.

Or were you thinking that someone could say `--target=depesz --post_url=x.x.x.x`, and then the `depesz.go` package would use the provided `post_url` instead of the default `http://explain.depesz.com/plan`?  If that's case, I can roll most of these changes back and just update to handle a custom `--post_url`

agneum replied:
Exactly, it would be cool to use a custom `post_url`

richyen replied:
I have added the post_url arg as an option

agneum commented on main.go:
```diff
@@ -13,11 +13,13 @@ import (
 
 	ctx := context.Background()
-	planner, err := visualizer.New(*target)
+	planner, err := visualizer.New(*target, *url, *key)
```
We could use a separate `Config` struct to pass parameters since there can be quite a lot of them.

richyen replied:
👍 -- I'll try creating a `config` struct and pass that in to the `visualizer` instantiator

richyen replied:
I haven't implemented this yet -- we are now at two args in the `New()` instantiator.  I can create the `Config` struct, but I wanted to confirm with you that you think this is the way to go.  Basically, 1) create a new package and import to `main.go` and `visualizer.go`, or 2) just pass two args?

agneum replied:
I think the first option is better: create a separate package for the `Config` structure and then pass it to the visualizer. So, an increase in the number of options will not lead to a change in the high-level API.

agneum commented on visualizer/visualizer.go:
```diff
@@ -3,24 +3,60 @@ package visualizer
+
+	case "dalibo":
+		return &Visualizer{
+			postURL: "https://explain.dalibo.com/new",
```
Using a constant has several advantages. This makes the code more robust and modifiable. I would rather keep the constants.

richyen replied:
Cool--I'm actually kind of new to golang, so I'm not very familiar with the advantages/disadvantages of using constants.  I'll roll back this change and re-introduce the constants

richyen replied:
FYI I have reverted the changes so the constants are back

agneum reviewed (changes requested):
@richyen I appreciate the contribution. That's an awesome job! I have some comments to improve the pull request.

richyen committed:
Support use of custom URL for visualizers
If users wish to deploy their own local copy of the 3 popular visualizers,
they can provide a custom URL to keep their plans (and schema) private

richyen committed:
merging conflicts

richyen committed:
Merge branch 'custom_explainer_1' into custom_explainer

* custom_explainer_1:
  Support use of custom URL for visualizers If users wish to deploy their own local copy of the 3 popular visualizers, they can provide a custom URL to keep their plans (and schema) private

richyen committed:
Update README

richyen committed:
Previous merge clobbered some imports.  Fixed and tested

agneum commented on visualizer/dalibo/dalibo.go:
```diff
@@ -11,24 +11,29 @@ import (
-func New() *Dalibo {
-	return &Dalibo{}
+func New(url string) *Dalibo {
+	if url == "" {
```
I suggest renaming the `url` variable (for example, to `postURL`) to avoid a collision with the imported package name `net / url`

agneum commented on README.md:
````diff
@@ -40,6 +40,7 @@ On default, make install puts the compiled binary in `go/bin`.
     postgres=# \o | plan-exporter
     ```
     * You may wish to specify `--target [dalibo|depesz|tensor]` to customize your visualizer
+    * You may also specify `--post_url <URL>` if you are deploying one of these targets locally
````
```suggestion
    * You may also specify `--post_url <URL>` if you are deploying one of these targets on-premise
```

agneum commented:
@richyen Impressive work! Just some minor changes are needed.

richyen committed:
rename vars to prevent collision with imported package

richyen committed:
consolidating commandline args into config for portability between packages

richyen committed:
Update README.md

Co-authored-by: Artyom Kartasov <agneum@users.noreply.github.com>

richyen committed:
lint and gofmt

richyen committed:
Merge branch 'custom_explainer' of github.com:richyen/plan-exporter into custom_explainer

* 'custom_explainer' of github.com:richyen/plan-exporter:
  Update README.md

NikolayS commented:
@richyen   👍  👍
